### PR TITLE
Make iterators AsyncIterable, Closes #89

### LINKED ISSUE #89: Feat: Make asynciterator asynciterable
In particular this logic https://github.com/LDflex/LDflex-Comunica/blob/f01ece6fa8cee720061efd5ddd23ed82921555e4/src/ComunicaEngine.ts#L158-L215 should be implemented as the `[Symbol.asyncIterator]` method of `AsyncIterator`

### PR DESCRIPTION
**Pull request by rubensworks**
While this will likely come at the cost of lower performance compared to plain data-events, we should add this for use cases where code readability is preferred over performance.

### CODE REVIEW

File: asynciterator.ts [hunk 1]
```diff
@@ -568,6 +567,47 @@ export class AsyncIterator<T> extends EventEmitter {
...
+              it.removeListener('error', reject);
+              resolve({ done: true, value: undefined });
+            }
+            else {
+              // In all other cases, wait for the iterator to become readable or ended
+              const nextCallback = () => {
+                it.removeListener('readable', nextCallback);
+                it.removeListener('end', nextCallback);
+                tryResolve();
+              };
```
Comment (jeswr) [inline asynciterator.ts]: You should be able to define this outside of `tryResolve` rather than re-defining it every time.

**Review by jeswr — APPROVED** [review]
LGTM (barring the minor optimisation above)

Note that this logic achieves the same effect over in LDflex https://github.com/LDflex/LDflex-Comunica/blob/f01ece6fa8cee720061efd5ddd23ed82921555e4/src/ComunicaEngine.ts#L161-L215.

I will leave @RubenVerborgh to merge as I do not have rights to make a subsequent npm release.

File: asynciterator.ts (hunk 1 above)
Comment (rubensworks) [reply asynciterator.ts]: Done!

**Comment by rubensworks** [comment]
@RubenVerborgh Could you have a look at this one? Will certainly make many comunica users very happy :-)

**Comment by RubenVerborgh** [comment]
@rubensworks So I first did f60ceccdc98f87dbafe2ceddcfc84de4e506a4d6 to minimize listener re-attachment within a single `next` iteration, but then I realize we could keep listeners attached in 43cedf106d1a6564682f3249e1a3c01ac867a19a, and then realized we _must_ do that to catch errors in between `next` calls. However, the test coverage is not complete for those cases yet.

**Comment by rubensworks** [comment]
I also had a `pendingError`-like approach in mind before, but I decided to not implement it like that before due to the chance of errors getting lost.
What could happen, is that the AsyncIterable is only being consumed partially, that an error is emitted after the last consumed element, and that the error will get lost because no further elements are getting consumed.
The alternative is that unhandled errors get thrown, but at least the errors do not get lost that way 😅 

So I guess we'll have to choose between either errors to always be explicitly forwarded (but them sometimes getting lost on partial consumption), or errors never getting lost, but them sometimes being unhandled errors.

----

After writing this, I just realized that we could also implement our own unhandled exception approach. If we notice (somehow) that the AsyncIterable is not consumed anymore, but we have a pending error, that we throw an unhandled exception. This might give us the best of both worlds.

**Comment by RubenVerborgh** [comment]
> After writing this, I just realized that we could also implement our own unhandled exception approach. If we notice (somehow) that the AsyncIterable is not consumed anymore, but we have a pending error, that we throw an unhandled exception. This might give us the best of both worlds.

I like the approach, except I'm not sure how to implement it with the ECMAScript `AsyncIterator` contract. How do I _know_ that no `next` will be called, if the error arrives between `next` calls?
What I am missing is a `destroy` method on ECMAScript `AsyncIterator`.

**Comment by rubensworks** [comment]
> How do I know that no next will be called, if the error arrives between next calls?

We may be able to do something by registering the created `AsyncIterator` into a [FinalizationRegistry](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/FinalizationRegistry), but it feels finicky...

**Comment by RubenVerborgh** [comment]
Ooof yes. Or perhaps for now at a comment that we recommend adding an error listener to the iterator if you don't plan on fully consuming it (since you can't detach a `for await` created thing yourself, because you don't have a reference to it).

_Or_ we could keep a list per iterator of the ESAsyncIterator instances created (or even allow only one), and then _destroy_ on the iterator also destroys all of them.

Actually, _destroy_ on the iterator might just work because it would also delete listeners.

**Comment by rubensworks** [comment]
@RubenVerborgh At last, I found some time to look into finalizing the unit tests and documenting the error events edge-case.

**Comment by RubenVerborgh** [comment]
That is quite amazing, thanks. So we good to go? Anything I should still do? (I see a review is pending, just checking if you need specific things.)

**Comment by rubensworks** [comment]
@RubenVerborgh Should be feature-complete AFAICS, so it should be ready to be merged if you agree :-)

File: asynciterator.ts [hunk 2]
```diff
@@ -611,7 +616,7 @@ export class AsyncIterator<T> extends EventEmitter implements AsyncIterable<T> {
         currentResolve = currentReject = pendingError = null;
         removeListeners();
       }
-      else if (pendingError !== null) {
+      else if (pendingError === null) {
```
Comment (RubenVerborgh) [inline asynciterator.ts:619]: 🤦 obvs I put that in on purpose to check your tests

**Comment by RubenVerborgh** [comment]
I was _so_ pleased that my code turned out to be the one in the end, only the see the dreadful `!==` error.
You go ahead and merge it, you did the tests and that's the hard work!

**Comment by rubensworks** [comment]
Merged! :-)

@RubenVerborgh, before you release, it might make sense to include #105 as well.

**Comment by jacoscaz** [comment]
@rubensworks @RubenVerborgh brilliant!!!

**Comment by rubensworks** [comment]
@RubenVerborgh Could you publish a new release? I'd like to include it in the next major release of Comunica, which I intend to release soon.

**Comment by RubenVerborgh** [comment]
@rubensworks Published v3.9.0! Thanks for your patience and continued support.
